File: src/models/Fee.js
const mongoose = require('mongoose');

const feeSchema = new mongoose.Schema({
  student: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Student is required']
  },
  feeType: {
    type: String,
    enum: ['room_rent', 'mess_fee', 'security_deposit', 'maintenance', 'electricity', 'water', 'internet', 'other'],
    required: [true, 'Fee type is required']
  },
  amount: {
    type: Number,
    required: [true, 'Amount is required'],
    min: [0, 'Amount cannot be negative']
  },
  dueDate: {
    type: Date,
    required: [true, 'Due date is required']
  },
  paidDate: Date,
  status: {
    type: String,
    enum: ['pending', 'paid', 'overdue', 'partial', 'waived'],
    default: 'pending'
  },
  paymentMethod: {
    type: String,
    enum: ['cash', 'card', 'upi', 'bank_transfer', 'cheque', 'online'],
    required: function() {
      return this.status === 'paid' || this.status === 'partial';
    }
  },
  transactionId: String,
  receiptNumber: String,
  month: {
    type: Number,
    min: 1,
    max: 12,
    required: [true, 'Month is required']
  },
  year: {
    type: Number,
    required: [true, 'Year is required']
  },
  description: String,
  lateFee: {
    type: Number,
    default: 0,
    min: [0, 'Late fee cannot be negative']
  },
  discount: {
    type: Number,
    default: 0,
    min: [0, 'Discount cannot be negative']
  },
  finalAmount: {
    type: Number,
    required: true
  },
  paidAmount: {
    type: Number,
    default: 0,
    min: [0, 'Paid amount cannot be negative']
  },
  balanceAmount: {
    type: Number,
    default: 0
  },
  paymentHistory: [{
    amount: Number,
    paidDate: Date,
    paymentMethod: String,
    transactionId: String,
    receiptNumber: String,
    paidBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    }
  }],
  reminders: [{
    sentDate: Date,
    type: {
      type: String,
      enum: ['email', 'sms', 'notification']
    },
    status: {
      type: String,
      enum: ['sent', 'delivered', 'failed']
    }
  }],
  room: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Room'
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  updatedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  notes: String
}, {
  timestamps: true
});

// Indexes for better query performance
feeSchema.index({ student: 1 });
feeSchema.index({ status: 1 });
feeSchema.index({ dueDate: 1 });
feeSchema.index({ month: 1, year: 1 });
feeSchema.index({ feeType: 1 });

// Pre-save middleware to calculate final amount and balance
feeSchema.pre('save', function(next) {
  // Calculate final amount
  this.finalAmount = this.amount + this.lateFee - this.discount;
  
  // Calculate balance amount
  this.balanceAmount = this.finalAmount - this.paidAmount;
  
  // Update status based on payment
  if (this.paidAmount === 0) {
    this.status = new Date() > this.dueDate ? 'overdue' : 'pending';
  } else if (this.paidAmount >= this.finalAmount) {
    this.status = 'paid';
    if (!this.paidDate) {
      this.paidDate = new Date();
    }
  } else {
    this.status = 'partial';
  }
  
  // Generate receipt number if paid
  if (this.status === 'paid' && !this.receiptNumber) {
    const year = new Date().getFullYear();
    const month = String(new Date().getMonth() + 1).padStart(2, '0');
    const random = Math.floor(Math.random() * 10000).toString().padStart(4, '0');
    this.receiptNumber = `RCP${year}${month}${random}`;
  }
  
  next();
});

// Method to add payment
feeSchema.methods.addPayment = function(paymentData) {
  this.paymentHistory.push({
    amount: paymentData.amount,
    paidDate: paymentData.paidDate || new Date(),
    paymentMethod: paymentData.paymentMethod,
    transactionId: paymentData.transactionId,
    receiptNumber: paymentData.receiptNumber,
    paidBy: paymentData.paidBy
  });
  
  this.paidAmount += paymentData.amount;
  this.paymentMethod = paymentData.paymentMethod;
  this.transactionId = paymentData.transactionId;
  
  return this.save();
};

// Method to add reminder
feeSchema.methods.addReminder = function(type, status = 'sent') {
  this.reminders.push({
    sentDate: new Date(),
    type: type,
    status: status
  });
  
  return this.save();
};

// Virtual for days overdue
feeSchema.virtual('daysOverdue').get(function() {
  if (this.status === 'overdue' || (this.status === 'pending' && new Date() > this.dueDate)) {
    const diffMs = new Date() - this.dueDate;
    return Math.floor(diffMs / (1000 * 60 * 60 * 24));
  }
  return 0;
});

// Virtual for payment percentage
feeSchema.virtual('paymentPercentage').get(function() {
  if (this.finalAmount === 0) return 100;
  return Math.round((this.paidAmount / this.finalAmount) * 100);
});

module.exports = mongoose.model('Fee', feeSchema);

File: src/main/java/com/hostel/model/Fee.java
package com.hostel.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;

import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import java.time.LocalDateTime;
import java.time.LocalDate;
import java.util.List;
import java.util.ArrayList;

@Document(collection = "fees")
public class Fee {
    @Id
    private String id;
    
    @NotNull(message = "Student ID is required")
    private String studentId;
    
    private String feeType; // room_rent, mess_fee, security_deposit, maintenance, electricity, water, internet, other
    
    @Min(value = 0, message = "Amount cannot be negative")
    private Double amount;
    
    @NotNull(message = "Due date is required")
    private LocalDate dueDate;
    
    private LocalDate paidDate;
    
    private String status = "PENDING"; // PENDING, PAID, OVERDUE, PARTIAL, WAIVED
    
    private String paymentMethod; // cash, card, upi, bank_transfer, cheque, online
    
    private String transactionId;
    
    private String receiptNumber;
    
    @Min(value = 1, message = "Month must be between 1 and 12")
    private Integer month;
    
    private Integer year;
    
    private String description;
    
    @Min(value = 0, message = "Late fee cannot be negative")
    private Double lateFee = 0.0;
    
    @Min(value = 0, message = "Discount cannot be negative")
    private Double discount = 0.0;
    
    private Double finalAmount;
    
    @Min(value = 0, message = "Paid amount cannot be negative")
    private Double paidAmount = 0.0;
    
    private Double balanceAmount = 0.0;
    
    private List<PaymentHistory> paymentHistory = new ArrayList<>();
    
    private List<Reminder> reminders = new ArrayList<>();
    
    private String roomId;
    
    private String createdBy;
    
    private String updatedBy;
    
    private String notes;
    
    @CreatedDate
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    private LocalDateTime updatedAt;
    
    // Constructors
    public Fee() {}
    
    public Fee(String studentId, String feeType, Double amount, LocalDate dueDate, Integer month, Integer year) {
        this.studentId = studentId;
        this.feeType = feeType;
        this.amount = amount;
        this.dueDate = dueDate;
        this.month = month;
        this.year = year;
        calculateFinalAmount();
    }
    
    public void calculateFinalAmount() {
        this.finalAmount = this.amount + this.lateFee - this.discount;
        this.balanceAmount = this.finalAmount - this.paidAmount;
        
        // Update status based on payment
        if (this.paidAmount == 0) {
            this.status = LocalDate.now().isAfter(this.dueDate) ? "OVERDUE" : "PENDING";
        } else if (this.paidAmount >= this.finalAmount) {
            this.status = "PAID";
            if (this.paidDate == null) {
                this.paidDate = LocalDate.now();
            }
        } else {
            this.status = "PARTIAL";
        }
    }
    
    // Getters and Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    
    public String getStudentId() { return studentId; }
    public void setStudentId(String studentId) { this.studentId = studentId; }
    
    public String getFeeType() { return feeType; }
    public void setFeeType(String feeType) { this.feeType = feeType; }
    
    public Double getAmount() { return amount; }
    public void setAmount(Double amount) { 
        this.amount = amount;
        calculateFinalAmount();
    }
    
    public LocalDate getDueDate() { return dueDate; }
    public void setDueDate(LocalDate dueDate) { this.dueDate = dueDate; }
    
    public LocalDate getPaidDate() { return paidDate; }
    public void setPaidDate(LocalDate paidDate) { this.paidDate = paidDate; }
    
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    
    public String getPaymentMethod() { return paymentMethod; }
    public void setPaymentMethod(String paymentMethod) { this.paymentMethod = paymentMethod; }
    
    public String getTransactionId() { return transactionId; }
    public void setTransactionId(String transactionId) { this.transactionId = transactionId; }
    
    public String getReceiptNumber() { return receiptNumber; }
    public void setReceiptNumber(String receiptNumber) { this.receiptNumber = receiptNumber; }
    
    public Integer getMonth() { return month; }
    public void setMonth(Integer month) { this.month = month; }
    
    public Integer getYear() { return year; }
    public void setYear(Integer year) { this.year = year; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public Double getLateFee() { return lateFee; }
    public void setLateFee(Double lateFee) { 
        this.lateFee = lateFee;
        calculateFinalAmount();
    }
    
    public Double getDiscount() { return discount; }
    public void setDiscount(Double discount) { 
        this.discount = discount;
        calculateFinalAmount();
    }
    
    public Double getFinalAmount() { return finalAmount; }
    public void setFinalAmount(Double finalAmount) { this.finalAmount = finalAmount; }
    
    public Double getPaidAmount() { return paidAmount; }
    public void setPaidAmount(Double paidAmount) { 
        this.paidAmount = paidAmount;
        calculateFinalAmount();
    }
    
    public Double getBalanceAmount() { return balanceAmount; }
    public void setBalanceAmount(Double balanceAmount) { this.balanceAmount = balanceAmount; }
    
    public List<PaymentHistory> getPaymentHistory() { return paymentHistory; }
    public void setPaymentHistory(List<PaymentHistory> paymentHistory) { this.paymentHistory = paymentHistory; }
    
    public List<Reminder> getReminders() { return reminders; }
    public void setReminders(List<Reminder> reminders) { this.reminders = reminders; }
    
    public String getRoomId() { return roomId; }
    public void setRoomId(String roomId) { this.roomId = roomId; }
    
    public String getCreatedBy() { return createdBy; }
    public void setCreatedBy(String createdBy) { this.createdBy = createdBy; }
    
    public String getUpdatedBy() { return updatedBy; }
    public void setUpdatedBy(String updatedBy) { this.updatedBy = updatedBy; }
    
    public String getNotes() { return notes; }
    public void setNotes(String notes) { this.notes = notes; }
    
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    
    public long getDaysOverdue() {
        if ("OVERDUE".equals(status) || ("PENDING".equals(status) && LocalDate.now().isAfter(dueDate))) {
            return LocalDate.now().toEpochDay() - dueDate.toEpochDay();
        }
        return 0;
    }
    
    public double getPaymentPercentage() {
        return finalAmount > 0 ? (paidAmount / finalAmount) * 100 : 100;
    }
    
    // Nested classes
    public static class PaymentHistory {
        private Double amount;
        private LocalDate paidDate;
        private String paymentMethod;
        private String transactionId;
        private String receiptNumber;
        private String paidBy;
        
        public PaymentHistory() {}
        
        public PaymentHistory(Double amount, String paymentMethod, String paidBy) {
            this.amount = amount;
            this.paymentMethod = paymentMethod;
            this.paidBy = paidBy;
            this.paidDate = LocalDate.now();
        }
        
        // Getters and Setters
        public Double getAmount() { return amount; }
        public void setAmount(Double amount) { this.amount = amount; }
        
        public LocalDate getPaidDate() { return paidDate; }
        public void setPaidDate(LocalDate paidDate) { this.paidDate = paidDate; }
        
        public String getPaymentMethod() { return paymentMethod; }
        public void setPaymentMethod(String paymentMethod) { this.paymentMethod = paymentMethod; }
        
        public String getTransactionId() { return transactionId; }
        public void setTransactionId(String transactionId) { this.transactionId = transactionId; }
        
        public String getReceiptNumber() { return receiptNumber; }
        public void setReceiptNumber(String receiptNumber) { this.receiptNumber = receiptNumber; }
        
        public String getPaidBy() { return paidBy; }
        public void setPaidBy(String paidBy) { this.paidBy = paidBy; }
    }
    
    public static class Reminder {
        private LocalDateTime sentDate;
        private String type; // email, sms, notification
        private String status; // sent, delivered, failed
        
        public Reminder() {}
        
        public Reminder(String type, String status) {
            this.type = type;
            this.status = status;
            this.sentDate = LocalDateTime.now();
        }
        
        // Getters and Setters
        public LocalDateTime getSentDate() { return sentDate; }
        public void setSentDate(LocalDateTime sentDate) { this.sentDate = sentDate; }
        
        public String getType() { return type; }
        public void setType(String type) { this.type = type; }
        
        public String getStatus() { return status; }
        public void setStatus(String status) { this.status = status; }
    }
}